src/Custom/Microsoft.Extensions.AI/OpenAIEmbeddingGenerator.cs

```diff
@@ -0,0 +1,109 @@
+    public OpenAIEmbeddingGenerator(EmbeddingClient embeddingClient, int? defaultModelDimensions = null)
+    {
+        Argument.AssertNotNull(embeddingClient, nameof(embeddingClient));
+        if (defaultModelDimensions < 1)
+        {
+            throw new ArgumentOutOfRangeException(nameof(defaultModelDimensions), "Value must be greater than 0.");
+        }
+
+        _embeddingClient = embeddingClient;
+        _dimensions = defaultModelDimensions;
+        _metadata = CreateMetadata("openai",
+            embeddingClient.Endpoint ?? OpenAIChatClient.DefaultOpenAIEndpoint,
+            embeddingClient.Model,
+            defaultModelDimensions);
+    }
+
+    /// <inheritdoc />
+    public async Task<GeneratedEmbeddings<Embedding<float>>> GenerateAsync(IEnumerable<string> values, EmbeddingGenerationOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        OpenAI.Embeddings.EmbeddingGenerationOptions? openAIOptions = ToOpenAIOptions(options);
+
+        var embeddings = (await _embeddingClient.GenerateEmbeddingsAsync(values, openAIOptions, cancellationToken).ConfigureAwait(false)).Value;
+
+        return new(embeddings.Select(e =>
+                new Embedding<float>(e.ToFloats())
+                {
+                    CreatedAt = DateTimeOffset.UtcNow,
+                    ModelId = embeddings.Model,
+                }))
+        {
+            Usage = new()
+            {
+                InputTokenCount = embeddings.Usage.InputTokenCount,
+                TotalTokenCount = embeddings.Usage.TotalTokenCount
+            },
+        };
+    }
+
+    /// <inheritdoc />
+    void IDisposable.Dispose()
+    {
+        // Nothing to dispose. Implementation required for the IEmbeddingGenerator interface.
+    }
+
+    /// <inheritdoc />
+    object? IEmbeddingGenerator.GetService(Type serviceType, object? serviceKey)
+    {
+        Argument.AssertNotNull(serviceType, nameof(serviceType));
+
+        return
+            serviceKey is not null ? null :
+            serviceType == typeof(EmbeddingGeneratorMetadata) ? _metadata :
+            serviceType == typeof(EmbeddingClient) ? _embeddingClient :
+            serviceType.IsInstanceOfType(this) ? this :
+            null;
+    }
+
+    /// <summary>Creates the <see cref="EmbeddingGeneratorMetadata"/> for this instance.</summary>
+    private static EmbeddingGeneratorMetadata CreateMetadata(string providerName, Uri providerUri, string? defaultModelId, int? defaultModelDimensions) =>
+        new(providerName, providerUri, defaultModelId, defaultModelDimensions);
+
+    /// <summary>Converts an extensions options instance to an OpenAI options instance.</summary>
+    private OpenAI.Embeddings.EmbeddingGenerationOptions? ToOpenAIOptions(EmbeddingGenerationOptions? options)
+    {
+        OpenAI.Embeddings.EmbeddingGenerationOptions openAIOptions = new()
+        {
+            Dimensions = options?.Dimensions ?? _dimensions,
+        };
```

Comment:
Ideally we'd follow the same pattern of getting an initial `EmbeddingGenerationOptions` from `AdditionalProperties` here as we will do for `OpenAIChatClient`, and eliminate L99-105 below.